MRG: revise documentation structure; add internals page. (#2184)
sourmash-bio/sourmash: doc/sourmash-internals.md

# A guide to the internals of sourmash

```{contents} Contents
:depth: 3
```

sourmash was created in 2015, and has been repeatedly reorganized,
refactored, and optimized to support ever larger databases, faster
queries, and new use cases. We've also regularly added new
functionality and features.  So sourmash can be pretty complicated
internally, and our user-facing documentation only covers a fraction
of its potential!

This document is a brain dump intended for expert users and sourmash
developers who want to understand how, why, and when to use various
sourmash features. It is unlikely ever to be comprehensive, so the
information you are interested in may not yet exist in this document,
but we are always happy to add to it -
[just ask in an issue!](https://github.com/sourmash-bio/sourmash/issues)

## Signatures and sketches

> sourmash operates on sketches. Each sketch is a collection of hashes.

Reviewer: ```suggestion
sourmash operates on sketches. Each sketch is a collection of hashes. Each hash is a whole number product of a k-mer and hash function. 
```
Maybe link the hash function to the ## making sketches section of the document. May wish to include another line about k-mers as well. Such as, 
"A k-mer is a short sub-sequence generated from a larger sequence file via a sliding window approach. 

I.e. If k=3, a 3-mer of an 8 base pair DNA molecule will have 6 individual 3-mers digested into hashes to populate a  sketch that are stored in a signature.

> The DNA sequence:
> AGTCATCG

> The k-mers of the sliding window process:
> [AGT].....
> .[GTC]....
> ..[TCA]...
> ...[CAT]..
> ....[ATC].
> .....[TCG]

> Hashes digested from each k-mer:
> 2118360698
> 1681319365
> 65865673
> 1255238040
> 364627471
> 659934874
"

Author: we could point at https://sourmash.readthedocs.io/en/latest/kmers-and-minhash.html for a basic intro, rather than rehashing (hah!) k-mers. What do you think?

Here are my revisions to the intro para - thoughts?

>sourmash operates on sketches. Each sketch is a collection of hashes,
which are in turn built from k-mers by appling a hash function
(currently always murmurhash) and a filtering function.  Each sketch
is contained in a signature wrapper that contains some metadata.

* a set of hashes;
* an optional abundance for each hash (when `track_abund` is True);
* a seed;
* a k-mer size;
* a molecule type;
* either a `num` (for MinHash) or a `scaled` value (for FracMinHash);

Signature objects (class `SourmashSignature`) contain a sketch (property `.minhash`) as well as additional information:
* an optional `name`
* an optional `filename`
* a license (currently must be CC0);
* an `md5sum(...)` method that returns a hash of the sketch.

For now, we tend to refer to signatures and sketches interchangeably,
because they are almost entirely 1:1 in the code base (but see [sourmash#616](https://github.com/sourmash-bio/sourmash/issues/616)).

The default signature interchange/serialization format is JSON, optionally
gzipped. This format is written and read by Rust code.

In general, a lot of effort in sourmash is spent managing collections of
signatures _before_ actually doing comparisons with them; see manifests,
and `Index` objects, below.

### Making sketches

Sketches are produced by hashing k-mers with murmurhash and then
keeping either the lowest `num` hashes (for MinHashes sketches) or

> keeping all hashes below `2**64 / scaled` (for FracMinHash sketches).

Reviewer: ```suggestion
keeping a selection of hashes below `2**64 / scaled` (for FracMinHash sketches).
```
Is this a more precise statement?

Author: Well, all hashes below `2**64/scaled` are kept, not a selection. The _selection_ is done by using the threshold of `2**64/scaled`, in other words.

Maybe add more detail - "keeping all hashes with values below `2**64/scaled`, which has the effect of selecting approximately one hash for every `scaled` k-mers - so, for a set of 100,000 k-mers, a scaled value of 1,000 would retain approximately 100 hashes."

whaddya think?

Reviewer: *Chef Kiss* I think your added detail clears up any ambiguity.

Author: updated in 91fbf7c1

The default MinHash sketches use parameters so that they are
compatible with mash sketches.

See [utils/compute-dna-mh-another-way.py](https://github.com/sourmash-bio/sourmash/blob/latest/utils/compute-dna-mh-another-way.py) for details on how k-mers are
hashed.

Note that if hashes are produced some other way (with a different hash
function) or from some source other than DNA, sourmash can still work
with them; only `sourmash sketch` does input checking.

### Compatibility checking

The point of the signatures and sketches is to enable certain kinds of

> rapid comparisons - Jaccard similarity and overlap analysis,

Reviewer: ```suggestion
rapid comparisons - Jaccard similarity, $JS(A,B) = \frac{|A \bigcap B|}{|A \bigcup B|}$ and overlap analysis,
```

Is overlap analysis the overlap coefficient, $OC(A,B) = \frac{|A \bigcap B|}{min(|A|,|B|)}$, or jaccard containment, $JC(A,B) = \frac{|A \bigcap B|}{|A|}$ .

Author: changed to:

>Jaccard similarity and number of overlapping k-mers,

which is kind of redundant (you can use # of overlapping k-mers to calculate Jaccard) but I'm ok with it since people are likely to have heard of Jaccard so it bridges to a concept they know.

true of sketches stored in indexed collections, e.g. SBTs; see [sourmash#1799](https://github.com/sourmash-bio/sourmash/issues/1799).)

In practice, sourmash does all necessary downsampling dynamically, but
returns the original sketches. This means that (for example) you can
do a low-resolution/high-scaled large scale search by specifying a
high scaled value, and then do a higher resolution comparison with
only the highly similar matches the results to do a more refined (see
below, Speeding up `gather` and `search`.)

### Num (MinHash) sketches support Jaccard similarity

"Regular" MinHash (or "num MinHash") sketches are implemented the same
way as in mash.  However, they are less well supported in sourmash,
because they don't offer the same opportunities for metagenome
analysis.  (See also [sourmash#1354](https://github.com/sourmash-bio/sourmash/issues/1354).)

Num MinHash sketches can always be compared by downsampling to a
common `num` value. This may need to be done manually using `sourmash
sig downsample`, however.

## K-mer sizes

There is no explicit restriction on k-mer sizes built into sourmash.

For highly specific genome and metagenome comparisons, we typically

> use k=21, k=31, or k=51. For a longer discussion, see [Assembly Free Analysis with K-mers](https://github.com/mblstamps/stamps2022/blob/main/kmers_and_sourmash/2022-stamps-assembly-free%20class.pdf) from STAMPS 2022.

Reviewer: ```suggestion
use k=21, k=31, or k=51. For a longer discussion, see [Assembly Free Analysis with K-mers](https://github.com/mblstamps/stamps2022/blob/main/kmers_and_sourmash/2022-stamps-assembly-free%20class.pdf) from STAMPS 2022 and a more general overview at [Using sourmash:a practical guide](https://sourmash.readthedocs.io/en/latest/using-sourmash-a-guide.html). 
```

Author: thanks! you can link directly within the docs by using Sphinx internal references - here, the reference would be `using-sourmash-a-guide.md`. I'm updating it in a separate commit at the moment ;)

Author: in fbbcee56

Reviewer: Interesting. Thanks for the info!

## Molecule types - DNA, protein, Dayhoff, and hydrophobic-polar

sourmash supports four different sequence encodings, which we refer to

> as "molecule: DNA (`--dna`), protein (`--protein`), Dayhoff,

Reviewer: ```suggestion
as molecule: DNA (`--dna`), protein (`--protein`), [Dayhoff](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-367/tables/1),
```
or https://en.wikipedia.org/wiki/Margaret_Oakley_Dayhoff#Table_of_Dayhoff_encoding_of_amino_acids

Author: thanks! using wikipedia link.

Author: in fbbcee56

> (`--dayhoff`), and hydrophobic-polar (`--hp`).

Reviewer: ```suggestion
(`--dayhoff`), and [hydrophobic-polar](https://sourmash.readthedocs.io/en/latest/sourmash-sketch.html#protein-encodings) (`--hp`).
```

Author: thanks - switching to internal ref.

Author: in fbbcee56

All FracMinHash sketches have exactly one molecule type, and can only
be compared to the same molecule type (and ksize).

DNA moltype sketches can be constructed from DNA input sequences using
`sourmash sketch dna`.

Protein, Dayhoff, and HP moltype sketches can be constructed from
protein input sequences using `sourmash sketch protein`, or from DNA
input sequences using `sourmash sketch translate`; `translate` will
translate in all six reading frames (see also
[orpheum](https://github.com/czbiohub/orpheum) from
[Botvinnik et al., 2021](https://www.biorxiv.org/content/10.1101/2021.07.09.450799v1)).
By default protein sketches will be created; dayhoff sketches can be
created by including `dayhoff` in the param string, e.g. `sourmash
sketch protein -p dayhoff`, and hydrophobic-polar sketches can be

> built with `hp` in the param string.

Reviewer: ```suggestion
built with `hp` in the param string, e.g. `sourmash sketch protein -p hp`.
```

rapid comparisons - Jaccard similarity and overlap analysis,
specifically. However, these comparisons can only be done between
compatible sketches.

Here, "compatible" means -
* the same MurmurHash seed (default 42);
* the same k-mer size/ksize (see k-mer sizes, below);
* the same molecule type (see molecule types, below);
* the same `num` or `scaled` (although see [this downsampling discussion](api-example.md#downsampling-signatures), and the next two sections);

sourmash uses selectors (`Index.select(...)`) to select sketches with
compatible ksizes, molecule types, and sketch types.

### Scaled (FracMinHash) sketches support similarity and containment

Per our discussion in [Irber et al., 2022](https://www.biorxiv.org/content/10.1101/2022.01.11.475838v2), FracMinHash sketches can always be compared
by downsampling to the max of the two scaled values.  (This is not always
true of sketches stored in indexed collections, e.g. SBTs; see [sourmash#1799](https://github.com/sourmash-bio/sourmash/issues/1799).)

In practice, sourmash does all necessary downsampling dynamically, but
returns the original sketches. This means that (for example) you can

> do a low-resolution/high-scaled large scale search by specifying a

Reviewer: Could "large scale" be changed to some synonym? Big, comprehensive, massive, sizable, huge?

```suggestion
do a low-resolution/high-scaled extensive search by specifying a
```

Author: updated in fbbcee56

> high scaled value, and then do a higher resolution comparison with

Reviewer: ```suggestion
high `scaled` value, and then do a higher resolution comparison with
```

> only the highly similar matches the results to do a more refined (see

Reviewer: ```suggestion
only the highly similar matches (see
```

Author: resolved in fbbcee56

below, Speeding up `gather` and `search`.)

### Num (MinHash) sketches support Jaccard similarity

"Regular" MinHash (or "num MinHash") sketches are implemented the same
way as in mash.  However, they are less well supported in sourmash,
because they don't offer the same opportunities for metagenome
analysis.  (See also [sourmash#1354](https://github.com/sourmash-bio/sourmash/issues/1354).)

> 

Reviewer: ```suggestion

### Conversion between Scaled (FracMinHash) and Num (MinHash) signatures with `downsample`

```
I think this could become its own section with the addition of scaled conversions.

Num MinHash sketches can always be compared by downsampling to a
common `num` value. This may need to be done manually using `sourmash
sig downsample`, however.

> 

Reviewer: ```suggestion

### Conversion between Scaled (FracMinHash) and Num (MinHash) signatures with `downsample`

As discussed in the previous sections, it is possible to adjust the `scaled` and `num` values to compare two FracMinHash signatures or two Num MinHash signatures. However, it is also possible to covert between the `scaled` and `num` signatures with the `sourmash sig downsample` command. For more details, review [command line](https://sourmash.readthedocs.io/en/latest/command-line.html#sourmash-signature-downsample-decrease-the-size-of-a-signature) document.

```
I think this would be a good spot for a section on scaled and num conversions because the two previous sections covered the conversion within the scaled and num domain, but the explicit discussion of converting between them is missing.

`sourmash sketch dna`.

Protein, Dayhoff, and HP moltype sketches can be constructed from
protein input sequences using `sourmash sketch protein`, or from DNA
input sequences using `sourmash sketch translate`; `translate` will
translate in all six reading frames (see also
[orpheum](https://github.com/czbiohub/orpheum) from
[Botvinnik et al., 2021](https://www.biorxiv.org/content/10.1101/2021.07.09.450799v1)).
By default protein sketches will be created; dayhoff sketches can be
created by including `dayhoff` in the param string, e.g. `sourmash
sketch protein -p dayhoff`, and hydrophobic-polar sketches can be
built with `hp` in the param string.

## Manifests

sourmash makes extensive use of signature manifests to support rapid
selection and lazy loading of signatures based on signature metadata
(name, ksize, moltype, etc.) See
[Blog post: Scaling sourmash to millions of samples](http://ivory.idyll.org/blog/2021-sourmash-scaling-to-millions.html)
for some of the motivation.

Manifests are an internal format that is not meant to be particularly
human readable, but the CSV format can be loaded into a spreadsheet
program if you're curious :).

> If in a zipfile (`.zip`) or SBT zipfile (`.sbt.zip`), manifests must

Reviewer: ```suggestion
If a collection of signatures are in a zipfile (`.zip`) or SBT zipfile (`.sbt.zip`), manifests must
```

Author: thanks! are => is, doing this in separate commit.

Author: updated in fbbcee56

be named `SOURMASH-MANIFEST.csv`. They can also be stored directly on
disk in CSV/gzipped CSV, or in a sqlite database; see
`sourmash sig manifest`, `sourmash sig check`, and `sourmash sig collect`
for manifest creation, management, and export utilities.

> Where signatures are stored individually in `Index` collections,

Reviewer: Is `Index` collections referring to SBT and LCA?
```suggestion
Where signatures are stored individually in `Index` collections (SBT and LCA),
```

Author: updated in fbbcee56

into memory (see `MultiIndex`, above) it is generally a bad idea to
use them - they may be slow to load and may consume a lot of
memory. They also do not support good loading error messages;
see [sourmash#1414](https://github.com/sourmash-bio/sourmash/issues/1414).

### Extensions for outputting index classes

Most commands that support saving signatures will save them in a
variety of formats, based on the extension provided (see
[sourmash#1890](https://github.com/sourmash-bio/sourmash/issues/1890)
for exceptions). The supported extensions are -

* `.zip` for `ZipFileLinearIndex`
* `.sqldb` for `SqliteIndex`
* `.sig` or `.sig.gz` for JSON/gzipped JSON
* `dirname/` to save in a directory hierarchy

The default signature save format is JSON, if the extension is not
recognized.

## Speeding up `gather` and `search`

There are two primary search commands in sourmash: `gather` and
`search`.

`gather` calculates a minimum metagenome cover as discussed in [Irber et al., 2022](https://www.biorxiv.org/content/10.1101/2022.01.11.475838v2). It
is mostly intended for querying a database with a metagenome, although
it can be used with genome queries, as well. It depends on overlap

> analyses and can only be used with FracMinHash sketches.

Reviewer: Is the overlap analysis containment or overlap coefficient(i.e. max containment)?

Author: neither, actually - number of shared bases (which can be used to calculate both containment index and max containment).

Author: committing:
>This approach relies on overlaps between genomes and metagenomes and can only be used with FracMinHash sketches.

Author: updated in fbbcee56

`search` typically is used to search a large database of sketches for
all similarity or containment matches above a threshold. Depending on
the query and the database, certain kinds of database indices may make
search much faster, especially when only a few matches are expected.

If you are doing many searches against the same database, indexing the
database as an SBT (with `sourmash index`) or as a `SqliteIndex`/sqldb
database is likely to provide a significant speed increase, albeit
with increased memory usage (SBT) or increased disk space (sqldb).

Conversely, `ZipFileLinearIndex` and the default `LCA_Database` are likely
to be poor choices for many searches - the former only supports linear
searches, and the latter needs to be loaded from disk and deserialized each
time.

### Running `gather` once

`gather` is typically used to search a metagenome against a large
database of sketches, as part of finding a minimum set cover. This can
be quite slow! Our current implementation (as of sourmash 4.1.0, [pull request sourmash#1370](https://github.com/sourmash-bio/sourmash/pull/1370)) does a single pass across the database to find all matches

> with an overlap above the provided threshold, and then organizes

Reviewer: ```suggestion
with a Jaccard similarity or containment above the provided threshold, and then organizes
```
